ddht/v5_1/pool.py

```diff
@@ -45,18 +46,23 @@ def __init__(
             events = Events()
         self._events = events
 
-        self._sessions = {}
+        self._sessions = LRU(128, callback=self._evict_session)
         self._sessions_by_endpoint = collections.defaultdict(set)
 
         self._outbound_packet_send_channel = outbound_envelope_send_channel
         self._inbound_message_send_channel = inbound_message_send_channel
 
+    def _evict_session(self, key: uuid.UUID, value: SessionAPI) -> None:
+        # callback to remove LRU evicted session from _sessions_by_endpoint
+        self._sessions_by_endpoint[value.remote_endpoint].remove(value)
+
     def remove_session(self, session_id: uuid.UUID) -> SessionAPI:
         try:
-            session = self._sessions.pop(session_id)
+            session = self._sessions[session_id]
         except KeyError:
             raise SessionNotFound(f"No session found with id: {session_id}")
 
+        del self._sessions[session_id]
```

Comment:
I think you can combine these into:

```python
try:
    del self._sessions[session_id]
except KeyError:
    raise SessionNotFound(...)
```

Reply:
I'm not sure, now with `self._sessions: LRUDict` - removing the element via `del` does not return the deleted element like it was previously with `pop()`. So we need two steps here - 1 to grab the `session` - 2 to delete the `session` from the `self._sessions: LRUDict`. Then we continue on the next line to remove the `session` from the `self.sessions_by_endpoint` store.

Reply:
right...  carry on.  Also, WTF is up with not supporting `pop()` in `LRU`?